physx/source/physxwebbindings/src/PxWebBindings.cpp

```diff
@@ -84,35 +164,52 @@ struct PxSimulationEventCallbackWrapper
   void onContact(const PxContactPairHeader &, const PxContactPair *pairs,
                  PxU32 nbPairs) {
     if (gContactPointsNeedClear) {
+      for (auto* point : gContactPoints) {
+        gContactPairPointPool.returnObj(point);
+      }
       gContactPoints.clear();
       gContactPointsNeedClear = false;
     }
     for (PxU32 i = 0; i < nbPairs; i++) {
       const PxContactPair &cp = pairs[i];
 
-      if (cp.flags & (PxContactPairFlag::eREMOVED_SHAPE_0 |
-                      PxContactPairFlag::eREMOVED_SHAPE_1))
+      if (cp.flags & (PxContactPairFlag::eREMOVED_SHAPE_0 | PxContactPairFlag::eREMOVED_SHAPE_1)) {
         continue;
+      }
 
-      std::vector<PxContactPairPoint> contactVec;
+      
       const PxU8 &contactCount = cp.contactCount;
       const PxU32 offset = gContactPoints.size();
       if (contactCount) {
-        contactVec.resize(contactCount);
-        pairs[i].extractContacts(&contactVec[0], contactCount);
-        gContactPoints.insert(gContactPoints.cend(), contactVec.cbegin(),
-                              contactVec.cend());
+        gContactPoints.resize(offset + contactCount);
+        for (PxU32 j = 0; j < contactCount; j++) {
+          if (gContactPoints[offset + j] == nullptr) {
+            gContactPoints[offset + j] = gContactPairPointPool.get();
+          }
+        }
+
+        gContactTemp.resize(contactCount);
+        pairs[i].extractContacts(gContactTemp.data(), contactCount);
+
+        // Copy the extracted contacts to the global contact points
+        for (PxU32 j = 0; j < contactCount; j++) {
+          *(gContactPoints[offset + j]) = gContactTemp[j];
+        }
       }
 
       if (cp.events & PxPairFlag::eNOTIFY_TOUCH_PERSISTS) {
-        call<void>("onContactPersist", cp.shapes[0], cp.shapes[1], contactCount,
-                   gContactPoints, offset);
+        // Raw pointer types passed to Javascript would be deleted after the call
+        // To avoid this, we need to override the 'raw_destructor' function.
+        // See the end of this file: 
+        //    template <> void raw_destructor<PxShape>(PxShape *) { /* do nothing */ }
+        //    template <> void raw_destructor<std::vector<PxContactPairPoint*>>(std::vector<PxContactPairPoint*>*) { }
+        //
+        call<void>("onContactPersist", cp.shapes[0], cp.shapes[1], contactCount, &gContactPoints, offset);
```

Comment:
Passing the adress `&gContactPoints` to avoid memory copy which also implements some performance.